doc/langref.html.in

```diff
@@ -6899,8 +6899,255 @@ fn foo() i32 {
 
       {#header_open|Result Location Semantics#}
       <p>
-      <a href="https://github.com/ziglang/zig/issues/2809">TODO add documentation for this</a>
+      During compilation, every Zig expression and sub-expression is assigned optional result location
+      information. This information dictates what type the expression should have (its result type), and
+      where the resulting value should be placed in memory (its result location). This is not an
+      implementation detail: this logic is codified into the Zig language specification. This system is
+      collectively referred to as "Result Location Semantics".
       </p>
+      {#header_open|Result Types#}
+      <p>
```

Comment:
The initial sentence above (line 6902) says the result location is "optional"?  Could you expand on that in this section?  (Or is it just the location that is optional, but not the type?)

Reply:
Also, if a "Result Type" is distinct from a "Type", that would be worth highlighting here.  The `*u32` example is very concrete and specific, but some of the examples below seem fuzzier to me (where the Result Type of an expression seems to be a range of types or type-like constraints, is not a specific Type).

Reply:
A result type, in general, must be an exact Zig type. The only one which is kind of fuzzy is `&e` with result type `[]T` requiring that `e` be "some array" of `T`. This is a bit of a white lie - internally, we actually preserve the detail that there's a slice involved, keeping types concrete. The details of this system are a bit tricky to explain precisely, so I opted for this slightly simplified version.

Reply:
> The initial sentence above (line 6902) says the result location is "optional"? Could you expand on that in this section?

To be honest, I thought the meaning of this was fairly clear? It's perfectly possible to write a statement like `_ = .{ 1, 2 }`, where there are no result types given, so the RHS sub-expression is given no type information. Does this need clarification?

Reply:
Ah, its optional when the context doesn't provide a constraint.  Make sense when you say it.  :)  Maybe just add something like "... optional (RLS is not defined when given no constraints as with `_` or `anytype`)"?

Reply:
I don't love the word "constraints" there, since it kind of implies RLS is more versatile than it is (see my response to your other comment). That said, I'll figure out how to tweak the wording here for clarity.